### ISSUE TITLE
Map-matching confidence value is misleading in some circumstances

### ISSUE THREAD

**Issue by danpat** [Project-OSRM/osrm-backend#3589]
I'm map-matching fairly infrequent GPS data (approx 2-3 minutes between coordinates in a relatively slow-moving vehicle) to roads in rural areas.  An example:

![screen shot 2017-01-20 at 01 27 05](https://cloud.githubusercontent.com/assets/1892250/22143649/984dadb4-deaf-11e6-94f0-cc3269553b4c.png)

The current map-matching plugin matches the trace just fine - however, the `confidence` calculation is misleading.  We base the `confidence` on the difference in the original trace length and the matched geometry length.  In the example above, the returned confidence was 0, although the match succeeded and the match geometry was as expected.

In the example above, there is little ambiguity about the trace.  The coordinates are close to the road, there are no other nearby roads that offer alternatives.  However, there is a significant difference in trace vs match lengths, due to the sparseness of the coordinate data.

We should see if we can improve the `confidence` measure in this situation somehow - what we're returning right now is somewhat misleading.

@TheMarex could we perhaps calculate the confidence based on which match candidates get used?  If the final route includes primarily the closest candidate phantom nodes, this seems like it would be  a pretty good indication of a good match.

**Comment by TheMarex** [comment]
@danpat yes we should switch to a scheme based on what is proposed in https://github.com/Project-OSRM/osrm-backend/pull/3531

We can probably train a classifier on (length_diff, number of feasible alternatives) for every trace point and collapse that to one value for the whole trace. This would be a great task for someone who wants to dig a little bit into some ML/statistics stuff.

**Comment by Nate-Wessel** [comment]
It would be great to see some more work on this issue. I had to do some pretty deep digging to figure out the meaning of the "confidence" value as things stand. I did finally find this:
https://www.mail-archive.com/osrm-talk@openstreetmap.org/msg00757.html
Which was helpful if not totally comforting. A more robust confidence measure would be greatly appreciated!

But maybe at the least someone should note in the docs that this value may be misleading under certain circumstances?
